### ISSUE TITLE
device.serial returns wrong data in samsung devices

### ISSUE THREAD

**Issue by RajanYugtia** [apache/cordova-plugin-device#91]
### Issue Type

- [x] Bug Report
- [ ] Feature Request
- [ ] Support Question

## Description
 I have added device plugin. Now i have added alert in device.serial. Now the original device serial and the alert information both are different. i don't know why but it's strange.

## Information
I have added the device plugin for getting device serial. Now i have added alert in constructor of page but it's giving different serial then the original phone serial. Please help me.

### Command or Code
alert("Serial number : "+this.device.serial)

### Environment, Platform, Device

Platform : android
Device :  samsung galaxt j4+ and  samsung galaxy S9+

### Version information

Operating System: 
samsung galaxy j4+ : 8.1.0 (Oreo)

cli packages:
    @ionic/cli-utils  : 1.19.2
    ionic (Ionic CLI) : 3.20.0

global packages:
    cordova (Cordova CLI) : 8.0.0

local packages:
    @ionic/app-scripts : 3.2.1
    Cordova Platforms  : android 7.0.0
    Ionic Framework    : ionic-angular 3.9.2

System:
    Android SDK Tools : 26.1.1
    Node              : v8.11.3
    npm               : 3.10.10
     OS                : Windows 8.1

## Checklist

- [x] I searched for already existing GitHub issues about this
- [ ] I updated all Cordova tooling to their most recent version
- [x] I included all the necessary information above

**Comment by HannaSanford** [comment]
I am having the same issue with Samsung Galaxy Tab A (2016) running android 8.1.0, samsung experience version 9.5

**Comment by janpio** [comment]
Can you both share the type of value/number you are getting?

**Comment by dwlrathod** [comment]
I am getting `ce061606e446672e02` from plugin when original is `RZ8H71ST0AX`.
You can find more details in [this](https://stackoverflow.com/a/18071939) answer.

**Comment by pankajabc10** [comment]
I am also getting problem with device.serial it shows undefined while another things like device.uuid, device.model etc are successfully accessed.

**Comment by emidizy** [comment]
@panjajabc10 were you able to find a resolution?  I m having same issue on android 9 devices. Device.serial reports "unknown" whereas other parameters like version, uuid and platform report correct values. My app is live and its using device.serial...

**Comment by HannaSanford** [comment]
I get numbers like `3300213f9f2aa58b` when the actual serial is something like`RF8KCxxxxxx`.

**Comment by thulari** [comment]
I don't get the plugin to work on a Huawei P20 lite...

**Comment by preejol** [comment]
@emidizy Have you found any solution ? as I am facing same problem for above android 8 devices

**Comment by emidizy** [comment]
@preejol  You may want to take a look at these references https://developer.android.com/about/versions/pie/android-9.0-changes-28

As from API 28, Google introduces some privacy controls to prevent access to device serial. It further classified it as a dangerous permission and therefore requires express  user consent each time you want to  access device serial. 
See https://developer.android.com/guide/topics/permissions/overview 

An alternative solution will be to generate a unique id for each of your users device and store it locally on the device say sqlite or something...

**Comment by gonzalovinas** [comment]
@emidizy great answer! But I could not find any ocurrence on the word "serial" in those links above you paste... Are those specific for this issue?

What about uuid ?

Thanks in advance!

**Comment by kschuster-systkom** [comment]
I tested the patch from @selmo47 but the system will thow an exception, see screenshot
![android-phone_state](https://user-images.githubusercontent.com/79266716/110318797-41657e00-800e-11eb-880a-062fea54bfd4.png)
So how i can get the serial number?

**Comment by breautek** [comment]
> So how i can get the serial number?

In API 28+ you need to qualify in these conditions:

https://developer.android.com/reference/android/os/Build#getSerial()

* If the calling app has been granted the `READ_PRIVILEGED_PHONE_STATE` permission; this is a privileged permission that can only be granted to apps preloaded on the device.
* If the calling app is the device or profile owner and has been granted the `READ_PHONE_STATE` permission. The profile owner is an app that owns a managed profile on the device; for more details see [Work profiles](https://developer.android.com/work/managed-profiles). Profile owner access is deprecated and will be removed in a future release.
* If the calling app has carrier privileges on any active subscription.
* If the calling app is the default SMS role holder

In other words, unless if you're a very special app, you cannot read the serial number anymore.

As of API 26, the native `serial` property will always be `unknown`.

On API 28 or lower, the native `getSerial` method will be `unknown` unless if your app matches one of the conditions above.
On API 29 or later, the native `getSerial` method will throw an exception unless if your app matches one of the conditions above.

I advise reading Google's [recommendations](https://developer.android.com/training/articles/user-data-ids) for what to use instead.

Closing this as won't fix.
